crates/configuration/src/shared/node.rs

```diff
@@ -32,124 +52,118 @@ impl From<(&str, &str)> for EnvVar {
     bootnodes_addresses: Vec<Multiaddr>,
-
-    /// Default resources. Override the default.
     resources: Option<Resources>,
-
-    /// Websocket port to use. Default to 9944 + n where n is the node index in the network (starting from 0).
     ws_port: Option<Port>,
-
-    /// RPC port to use. Default to 9933 + n where n is the node index in the network (starting from 0).
-    // [TODO]: start at a different default to avoid overlap between ws_port and rpc_port when node count >= 12 ?
     rpc_port: Option<Port>,
-
-    /// Prometheus port to use. Default to 9615 + n where n is the node index in the network (starting from 0).
     prometheus_port: Option<Port>,
-
-    /// P2P port to use. Default to 30333 + n where n is the node index in the network (starting from 0)
     p2p_port: Option<Port>,
-
-    /// libp2p cert hash to use with `webrtc` transport.
     p2p_cert_hash: Option<String>,
-
-    /// Database snapshot. Override the default.
     db_snapshot: Option<AssetLocation>,
 }
 
 impl NodeConfig {
+    /// Node name (should be unique).
+    // TODO: handle unique in validation
     pub fn name(&self) -> &str {
         &self.name
     }
 
+    /// Image to run (only podman/k8s). Override the default.
+    // TODO: missing test of default overridding
     pub fn image(&self) -> Option<&Image> {
         self.image.as_ref()
     }
 
+    /// Command to run the node. Override the default.
+    // TODO: missing test of default overriding
     pub fn command(&self) -> Option<&Command> {
         self.command.as_ref()
     }
 
+    /// Arguments to use for node. Appended to default.
+    // TODO: missing test of default appending
     pub fn args(&self) -> Vec<&Arg> {
         self.args.iter().collect()
     }
 
+    /// Wether the node is a validator.
     pub fn is_validator(&self) -> bool {
         self.is_validator
     }
 
+    /// Whether the node keys must be added to invulnerables.
     pub fn is_invulnerable(&self) -> bool {
         self.is_invulnerable
     }
 
+    /// Whether the node is a bootnode.
     pub fn is_bootnode(&self) -> bool {
         self.is_bootnode
     }
 
+    /// Node initial balance present in genesis.
     pub fn initial_balance(&self) -> u128 {
         self.initial_balance
     }
 
+    /// Environment variables to set (inside pod for podman/k8s, inside shell for native).
     pub fn env(&self) -> Vec<&EnvVar> {
         self.env.iter().collect()
     }
 
+    /// List of node's bootnodes addresses to use. Appended to default.
+    // TODO: missing test appending to default.
     pub fn bootnodes_addresses(&self) -> Vec<&Multiaddr> {
         self.bootnodes_addresses.iter().collect()
     }
 
+    /// Default resources. Override the default.
+    // TODO: missing test default override.
     pub fn resources(&self) -> Option<&Resources> {
         self.resources.as_ref()
     }
 
+    /// Websocket port to use. Default to 9944 + n where n is the node index in the network (starting from 0).
```

Comment:
This is not the expected behavior, the current version assign a `random` port if not set.

Reply:
Alright, but in the case of the config validation, do we want to returns an error if there are 2 time the same ports ? Because actually there is no validation for port uniqueness

Reply:
The strategy that I described in the comment was more like an idea

Reply:
Ahh yes, we should track the ports and ensure those are unique.

Reply:
Should they be unique across the whole config ? For example if a collator has port 9987, a node can't have the same port, right ?

Reply:
You are right, since those are the ports mapped to the local machine so should need to be unique.

Reply:
Should we use incremental port numbers? Meaning that relay chain will start from 9615 and increment for each node... then parachains will start at 9100 -> 9500 but incremental. This way even though ports are random we ensure that there will not be any re usability. In addition, we could maybe just "keep" in a struct all the init'ed ports, but Im not sure what happens with parallel processes  (thinking out loud here)

Reply:
Yes @wirednkod this is kind of what I was trying to propose but at some point with really big networks we could have overlap ? Imho, passing a hashmap/array of port could be painful in term of code but is doable, the other thing I was thinking about is about having random ports, do we really want to do something for the user (remplace a duplicate with a random port) when we could just show an error saying some port are duplicate and let the user change them in the config ? @pepoviola wdyt ?

Reply:
I think that if the user set `dup` ports in the config we should show the error and let the user to resolve.

Reply:
Ok, nice, I think this is the simplest way